docs/zkapps/tutorials/05-common-types-and-functions.mdx

```diff
@@ -0,0 +1,358 @@
+
+Merkle trees are currently in the `Experimental` module of SnarkyJS. So start off by importing `Experimental`:
+```ts
+import {
+  ...
+  Experimental,
+  ...
+} from 'snarkyjs'
+```
+
+When using Merkle trees, you initialize an instance of the data structure on the "ui" side of your application at a particular height. 
+
+In the smart contract though, you need to provide a "witness" for each update to the tree, which proves the connection between the existing tree in the smart contract, and the updated tree.
+
+Here is a simple example of a contract that stores the root of a merkle tree, where each leaf stores a number, and the smart contract has an `update` function that adds a number to the leaf, while checking that the number added was less than 10:
+
+```ts
+...
+  @state(Field) treeRoot = State<Field>();
+...
+  @method init(initialRoot: Field) {
+    this.treeRoot.set(initialRoot);
+  }
+
+  @method update(
+    leafWitness: MerkleWitness20, 
+    numberBefore: Field,
+    incrementAmount: Field,
+  ) {
+    const initialRoot = this.treeRoot.get();
+    this.treeRoot.assertEquals(initialRoot);
+
+    incrementAmount.assertLt(Field(10));
+
+    // check the initial state matches what we expect
+    const rootBefore = leafWitness.calculateRoot(numberBefore);
+    rootBefore.assertEquals(initialRoot);
+
+    // compute the root after incrementing
+    const rootAfter = leafWitness.calculateRoot(numberBefore.add(incrementAmount));
+
+    // set the new root
+    this.treeRoot.set(rootAfter);
+  }
+```
+
+And code to interact with it:
+
+```ts
+    // initialize the zkapp
+    const zkapp = new BasicMerkleTreeContract(basicTreeZkAppAddress);
+
+    // create a new tree
+    const height = 20;
+    const tree = new Experimental.MerkleTree(height);
+    class MerkleWitness extends Experimental.MerkleWitness(height) {}
+
+    // deploy the smart contract
+    const deploy_txn = await Mina.transaction(deployerAccount, () => {
+      AccountUpdate.fundNewAccount(deployerAccount);
+      zkapp.deploy({ zkappKey: basicTreeZkAppPrivateKey });
+      // get the root of the new tree to use as the initial tree root
+      zkapp.init(tree.getRoot());
+      zkapp.sign(basicTreeZkAppPrivateKey);
+    });
+    await deploy_txn.send().wait();
+
+    const incrementIndex = 522;
```

Comment:
nit: the code below would become slightly cleaner if we instantiate incrementIndex as a bigint literal here: 
```ts
const incrementIndex = 522n;
```

Reply:
We currently target ES2019 with zk project - so the bignum notation doesn't work unfortunately - do think we should consider updating that tho to unlock newer syntax like this

Reply:
oh I see -- yes I'll make a note to update the target!